vllm/model_executor/layers/quantization/gptq.py

```diff
@@ -0,0 +1,226 @@
+        ]
+
+    @classmethod
+    def from_config(cls, config: Dict[str, Any]) -> "GPTQConfig":
+        weight_bits = cls.get_from_keys(config, ["bits"])
+        group_size = cls.get_from_keys(config, ["group_size"])
+        desc_act = cls.get_from_keys(config, ["desc_act"])
+        return cls(weight_bits, group_size, desc_act)
+
+    def get_linear_method(self) -> "GPTQLinearMethod":
+        return GPTQLinearMethod(self)
+
+    def get_scaled_act_names(self) -> List[str]:
+        return []
+
+
+class GPTQLinearMethod(LinearMethodBase):
+    """Linear method for GPTQ.
+
+    Args:
+        quant_config: The GPTQ quantization config.
+    """
+
+    def __init__(self, quant_config: GPTQConfig):
+        self.quant_config = quant_config
+        self.use_exllama = True
+
+    def create_weights(self, input_size: int, output_size: int,
+                       params_dtype: torch.dtype,
+                       parallel_type: str = "none") -> Dict[str, torch.Tensor]:
+        if input_size % self.quant_config.group_size != 0:
+            raise ValueError(
+                "The input size is not aligned with the quantized "
+                "weight shape. This can be caused by too large "
+                "tensor parallel size.")
+        if output_size % self.quant_config.pack_factor != 0:
+            raise ValueError(
+                "The output size is not aligned with the quantized "
+                "weight shape. This can be caused by too large "
+                "tensor parallel size.")
+
+        qweight = Parameter(
+            torch.empty(
+                input_size // self.quant_config.pack_factor,
+                output_size,
+                device="cuda",
+                dtype=torch.int32,
+            ),
+            requires_grad=False,
+        )
+        set_weight_attrs(
+            qweight, {
+                "input_dim": 0,
+                "output_dim": 1,
+                "packed_dim": 0,
+                "pack_factor": self.quant_config.pack_factor,
+            })
+        g_idx = Parameter(
+            torch.tensor(
+                [i // self.quant_config.group_size for i in range(input_size)],
+                device="cuda",
+                dtype=torch.int32,
+            ),
+            requires_grad=False,
+        )
+        set_weight_attrs(g_idx, {"input_dim": 0})
+        tp_size = get_tensor_model_parallel_world_size()
+        if parallel_type == "row" and tp_size > 1 and (self.quant_config.desc_act
+                and self.quant_config.group_size != -1):
+            input_size = input_size * tp_size
+            use_exllama = Parameter(torch.tensor(False, dtype=torch.bool, device="cuda"), requires_grad=False)
+        else:
+            use_exllama = Parameter(torch.tensor(True, dtype=torch.bool, device="cuda"), requires_grad=False)
+        if self.quant_config.desc_act or self.quant_config.group_size == -1:
+            input_dim = None
+        else:
+            input_dim = 0
```

Comment:
Can you explain the logic here?

Reply:
Exllama reorders the features of the linear weight as well as the input matrix for desc-act models. However for row parallel layers, the input is sharded into different partitions and we cannot reorder it as a whole without extra gathering, in which case an alternative kernel is used instead.
Currently the scales and qzeros are sharded for Exllama kernel while fully copied for the other, following their original implementation. Probably this could be simplified by using fully copy for both and I'll give it a try.